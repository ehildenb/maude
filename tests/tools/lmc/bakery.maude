set show timing off .

load ../../../contrib/systems/bakery.maude
load ../../../contrib/tools/lmc/lgraph-search.maude

fmod RENAMED-FVP-NARROWING-GRAPH is
   protecting ( FVP-NARROWING-GRAPH + GRAPH-ANALYSIS )
            * ( sort   Nat to   NatP
              , sort NzNat to NzNatP
              ) .
endfm

--- Module BAKERY-NARROWING-GRAPH
--- =============================

mod BAKERY-NARROWING-GRAPH is
   protecting BAKERY .
    extending RENAMED-FVP-NARROWING-GRAPH .

    eq #M = upModule('BAKERY, true) .

    op upIdle : -> [Term] .
    -----------------------
    eq upIdle = upTerm([idle]) .

    op init : NatP -> [Term] .
    --------------------------
    eq init((0).NatP) = upTerm(0 ; 0 ; [idle] [idle]) .
    eq init((1).NatP) = '_;_;_['X:Name,'X:Name,'__[upIdle,upIdle]] .
    eq init((2).NatP) = '_;_;_['X:Name,'Y:Name,'__[upIdle,'PWS:ProcWaitSet]] .
    eq init((3).NatP) = '_;_;_['__['s.Name,'X:Name],'Y:Name,'__['`[_`]['crit['__['s.Name,'X:Name]]], upIdle, upIdle, 'PWS:ProcWaitSet]] .
    eq init((4).NatP) = '_;_;_['X:Name,'Y:Name,'__[upIdle,upIdle]] .
endm

--- Low-level functions
--- -------------------

--- ### ensure that init(0..4) are well formed

reduce wellFormed(#M, init(0)) .
reduce wellFormed(#M, init(1)) .
reduce wellFormed(#M, init(2)) .
reduce wellFormed(#M, init(3)) .
reduce wellFormed(#M, init(4)) .

--- ### Function `step`

reduce step(state(init(0))) .
reduce step(state(init(1))) .
reduce step(state(init(2))) .
reduce step(state(init(3))) .

--- ### Function `fold`

reduce fold ( state('_;_;_['__['s.Name,'V1:Name],'__['s.Name,'V1:Name],'__['`[_`]['idle.ModeIdle],'`[_`]['idle.ModeIdle]]])
            , state('_;_;_['X:Name,              'Y:Name,              '__['`[_`]['idle.ModeIdle],'`[_`]['idle.ModeIdle]]])
            ) .
reduce fold(state(init(1)), state(init(2))) .
reduce fold(state(init(1)), state(init(4))) .

reduce not (fold(state(init(4)), state(init(3))) :: Fold) .

--- ### Function `bfs`

reduce bfs(state(init(0)), 0) .
reduce bfs(state(init(1)), 0) .
reduce bfs(state(init(2)), 0) .
reduce bfs(state(init(3)), 0) .

reduce bfs(state(init(0)), 2) .
reduce bfs(state(init(1)), 2) .
reduce bfs(state(init(2)), 2) .
reduce bfs(state(init(3)), 2) .

--- achieves finite state graph
reduce bfs(state(init(1))) .

reduce bfs(state(init(4)), 10) .

--- Module BAKERY-FVP-CTOR-NARROWING-GRAPH
--- ======================================

mod BAKERY-FVP-CTOR-NARROWING-GRAPH is
   protecting BAKERY-FVP-CTOR .
    extending RENAMED-FVP-NARROWING-GRAPH .

    var N : NatP . vars NeNS NeNS' : NeNodeSet .

    op states : NodeSet -> [NodeSet] .
    ----------------------------------
    eq states(N) = state(init(N)) .

    eq states(.NodeSet)     = .NodeSet .
    eq states(NeNS ; NeNS') = states(NeNS) ; states(NeNS') .

    eq #M = upModule('BAKERY-FVP-CTOR, true) .

    op init : NatP -> [Term] .
    --------------------------
    eq init(0) = '<_`,_`,_`,_>['sleep.NcMode ,'_+_['1.NzNat,'1.NzNat,'1.NzNat],'wait.NcMode,'_+_['1.NzNat,'1.NzNat]] .
    eq init(1) = '<_`,_`,_`,_>[ 'PNC:NcMode   , 'X1:Nat   , 'QC:Mode      , 'X2:Nat   ] .
    eq init(2) = '<_`,_`,_`,_>[ 'PC:Mode      , 'Y1:Nat   , 'QNC:NcMode   , 'Y2:Nat   ] .
    eq init(3) = '<_`,_`,_`,_>[ 'sleep.NcMode , 'X0:Nat   , 'sleep.NcMode , 'X0:Nat   ] .
    eq init(4) = '<_`,_`,_`,_>[ 'sleep.NcMode , 'X0:Nat   , 'sleep.NcMode , 'X1:Nat   ] .
    eq init(5) = '<_`,_`,_`,_>[ 'PNC:NcMode   , 'X1:Nat   , 'QC:Mode      , 'X2:NzNat ] .
    eq init(6) = '<_`,_`,_`,_>[ 'PC:Mode      , 'Y1:NzNat , 'QNC:NcMode   , 'Y2:Nat   ] .
    eq init(7) = '<_`,_`,_`,_>[ 'wait.NcMode  , 'Y1:Nat   , 'QC:Mode      , 'Y2:NzNat ] .
    eq init(8) = '<_`,_`,_`,_>[ 'QC:Mode      , 'Y1:NzNat , 'wait.NcMode  , 'Y2:Nat   ] .
    eq init(9) = '<_`,_`,_`,_>[ 'crit.Mode    , 'X2:Nat   , 'crit.Mode    , 'X3:Nat   ] .
endm

***(Desired initial states and invariant:
   vars X0 X Y X' Y' : Nat .

Set of initial states:  < sleep, X0, sleep, X0 > .

  vars P Q : Mode .    vars P' Q' : NcMode .  vars

Mutex Invariant:  < P', X, Q, Y > \/ < P, X', Q', Y' > .
             or: state(init(1)) \/ state(init(2)) .
)


reduce wellFormed(#M, init(0)) .
reduce wellFormed(#M, init(1)) .
reduce wellFormed(#M, init(2)) .
reduce wellFormed(#M, init(3)) .
reduce wellFormed(#M, init(4)) .
reduce wellFormed(#M, init(5)) .
reduce wellFormed(#M, init(6)) .
reduce wellFormed(#M, init(7)) .
reduce wellFormed(#M, init(8)) .
reduce wellFormed(#M, init(9)) .

reduce step(states(0)) .
reduce step(states(1)) .
reduce step(states(2)) .
reduce step(states(3)) .

reduce states(3) <= states(1 ; 2) .
reduce states(1) <= states(1 ; 2) .
reduce states(2) <= states(1) .

reduce state('wait.NcMode)           <= state('PC:Mode) .
reduce state('_+_['1.NzNat,'V3:Nat]) <= state('Y1:NzNat) .
reduce state('V2:NcMode)             <= state('QNC:NcMode) .
reduce state('V3:Nat)                <= state('Y2:Nat) .

--- TODO: It would be nice to be able to prove the following
reduce state('Y1:NzNat) <= state('_+_['1.NzNat,'V3:Nat]) .

reduce state('<_`,_`,_`,_>[ 'wait.NcMode , '_+_['1.NzNat,'V3:Nat] ,  'V2:NcMode , 'V3:Nat ])
    <= state('<_`,_`,_`,_>[ 'PC:Mode     , 'Y1:NzNat              , 'QNC:NcMode , 'Y2:Nat ]) .


reduce state('sleep.NcMode) <= state('PNC:NcMode) .
reduce state('0.Nat)        <= state('X1:Nat) .
reduce state('V2:NcMode)    <= state('QC:Mode) .
reduce state('V3:Nat)       <= state('X2:NzNat) .

reduce state('<_`,_`,_`,_>[ 'sleep.NcMode , '0.Nat  , 'V2:NcMode , 'V3:Nat   ])
    <= state('<_`,_`,_`,_>[ 'PNC:NcMode   , 'X1:Nat , 'QC:Mode   , 'X2:NzNat ]) .

reduce   state('<_`,_`,_`,_>[ 'sleep.NcMode , '0.Nat    , 'V2:NcMode    , 'V3:Nat   ])
    <= ( state('<_`,_`,_`,_>[ 'PC:Mode      , 'Y1:NzNat , 'QNC:NcMode   , 'Y2:Nat   ])
       ; state('<_`,_`,_`,_>[ 'PNC:NcMode   , 'X1:Nat   , 'QC:Mode      , 'X2:NzNat ])
       ; state('<_`,_`,_`,_>[ 'sleep.NcMode , '0.Nat    , 'sleep.NcMode , '0.Nat    ])
       ) .

reduce intersect(state('V2:NzNat), state('V3:Nat)) .
reduce intersect(state('V2:NzNat), state('V3:Nat)) .
reduce intersect( state('<_`,_`,_`,_>['V1:Mode,'0.Nat,'crit.Mode,'V2:Nat])
                , state('<_`,_`,_`,_>['V1:Mode,'_+_['V2:Nat,'V3:NzNat],'crit.Mode,'V2:Nat])
                ) .

--- both initial states yield finite state-graphs
reduce bfs(states(1)) .
reduce bfs(states(2)) .
reduce bfs(states(3)) .
reduce bfs(states(1 ; 2)) .

reduce nodes(bfs(states(1))) .
reduce nodes(bfs(states(2))) .
reduce nodes(bfs(states(3))) .
reduce nodes(bfs(states(1 ; 2))) .

reduce nodes(bfs(states(1 ; 2))) <= (nodes(bfs(states(1))) ; nodes(bfs(states(2)))) .
reduce (nodes(bfs(states(1))) ; nodes(bfs(states(2)))) <= nodes(bfs(states(1 ; 2))) .

reduce states(3)                            <= (states(1 ; 2)) .
reduce states(3)                            <= nodes(bfs(states(1 ; 2))) .
reduce (states(1 ; 2))                      <= nodes(extend(states(1 ; 2))) .
reduce (states(1 ; 2))                      <= nodes(bfs(states(1 ; 2))) .
reduce nodes(extend(states(1 ; 2)))         <= states(1 ; 2) .
reduce nodes(extend(states(1 ; 2)))         <= nodes(bfs(states(1 ; 2))) .
reduce nodes(extend(states(5 ; 6 ; 7 ; 8))) <= states(5 ; 6 ; 7 ; 8) .

--- state init(3) stays within init(1) ; init(2)
reduce nodes(bfs(states(3))) <= (states(1 ; 2)) .

--- even though in general init(1) ; init(2) is *not* an invariant
reduce invariant(states(1 ; 2)) .
reduce invariant(states(4 ; 5 ; 6)) .
reduce invariant(states(5 ; 6 ; 7 ; 8)) .
reduce nodes(bfs(states(1 ; 2))) <= (states(1 ; 2)) .
reduce check states(3) stable in states(1 ; 2) .

--- state 3 never reaches the bad state 9, but states 1 ; 2 can
reduce states(3)     =>* states(9) .
reduce states(1 ; 2) =>* states(9) .
