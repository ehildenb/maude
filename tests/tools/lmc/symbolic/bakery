#!/bin/sh

MAUDE_LIB=$srcdir/../../../src/Main
export MAUDE_LIB

../../../src/Main/maude \
  < $srcdir/bakery.maude -no-banner -no-advise \
  > bakery.out 2>&1

diff $srcdir/bakery.expected bakery.out > /dev/null 2>&1
