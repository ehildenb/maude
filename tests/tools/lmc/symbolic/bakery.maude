
--- Lamport's Bakery protocol in Maude.
--- From: All About Maude - A High-Performance Logical Framework. 
fmod BAKERY-SYNTAX is 
  sort Name .
  op 0 : -> Name [ctor] .
  op s : -> Name [ctor] .
  op __ : Name Name -> Name [ctor comm assoc id: 0] .

  sorts ModeIdle ModeWait ModeCrit Mode . 
  subsorts ModeIdle ModeWait ModeCrit < Mode . 
  sorts ProcIdle ProcWait Proc ProcIdleSet ProcWaitSet ProcSet . 
  subsorts ProcIdle < ProcIdleSet . 
  subsorts ProcWait < ProcWaitSet . 
  subsorts ProcIdle ProcWait < Proc < ProcSet . 
  subsorts ProcIdleSet < ProcWaitSet < ProcSet . 
  
  op idle : -> ModeIdle . 
  op wait : Name -> ModeWait . 
  op crit : Name -> ModeCrit . 
  op `[_`] : ModeIdle -> ProcIdle . 
  op `[_`] : ModeWait -> ProcWait . 
  op `[_`] : Mode -> Proc . 
  op none : -> ProcIdleSet . 
  op __ : ProcIdleSet ProcIdleSet -> ProcIdleSet [assoc comm id: none] . 
  op __ : ProcWaitSet ProcWaitSet -> ProcWaitSet [assoc comm id: none] .
  op __ : ProcSet ProcSet -> ProcSet [assoc comm id: none] .

  sort Conf . 
  op _;_;_ : Name Name ProcSet -> Conf .
endfm


mod BAKERY is
  protecting BAKERY-SYNTAX . 
  
  var PS : ProcSet .  vars N M : Name . 
  
  rl [wake] : N ; M ; [idle] PS    => s N ; M ; [wait(N)] PS .
  rl [crit] : N ; M ; [wait(M)] PS => N ; M ; [crit(M)] PS .
  rl [exit] : N ; M ; [crit(M)] PS => N ; s M ; [idle] PS .
endm


load symbolic-checker

--- prop definitions for bakery
(mod BAKERY-SATISFACTION1 is
  pr BAKERY .
  pr SYMBOLIC-CHECKER .

  subsort Conf < State .

  ops ever-wait? ever-crit? : -> Prop . 

  vars N M : Name . vars PS : ProcSet . 
  
  eq s N ; M ; PS |= ever-wait? = true [variant] . 
  eq 0 ; M ; PS    |= ever-wait? = false [variant] . 
  eq N ; s M ; PS |= ever-crit? = true [variant] .
  eq N ; 0 ; PS    |= ever-crit? = false [variant] . 
endm)

set verbose on .

*** true
(lfmc N ; N ; [idle] [idle] |= [] (ever-wait? -> <> ever-crit?) .)

*** deadlock counterexample
(lfmc N ; M ; IS:ProcIdleSet |= [] (ever-wait? -> <> ever-crit?) .)

*** spurious counterexample
(lfmc N ; M ; WS:ProcWaitSet |= [] (ever-wait? -> <> ever-crit?) .)


(mod BAKERY-SATISFACTION2 is
  pr BAKERY .
  pr SYMBOLIC-CHECKER .

  subsort Conf < State .

  ops ex? ini : -> Prop . 

  var WS : ProcWaitSet . var IS : ProcIdleSet . var PS : ProcSet . 
  vars N M M1 M2 : Name .

--- NOTE: ProcIdleSet < ProcWaitSet
  eq N ; M ; WS |= ex? = true [variant] .
  eq N ; M ; [crit(M1)] WS |= ex? = true [variant] .
  eq N ; M ; [crit(M1)] [crit(M2)] PS |= ex? = false [variant] .
endm)

*** no counterexample
(lmc [10] N:Name ; N:Name ; [idle] [idle] |= [] ex? .)

*** true
(lfmc N:Name ; N:Name ; [idle] [idle] |= [] ex? .)

*** false
(lfmc N ; M ; WS |= [] ex?  .)

*** false (infinite graph, but a counterexample exists)
(lfmc N ; N ; WS |= [] ex?  .)

(mod BAKERY-SATISFACTION2-ABS is
  pr BAKERY-SATISFACTION2 .

  var PS : ProcSet .  vars N M L : Name . var IS : ProcIdleSet .

--- bisimilar equational abstraction
  eq s s s N M ; M ; PS [wait(s N M)] [wait(s s N M)]
   =   s s N M ; M ; PS [wait(s N M)]
  [variant] .
endm)

*** true
(lfmc N ; N ; IS |= [] ex?  .)

