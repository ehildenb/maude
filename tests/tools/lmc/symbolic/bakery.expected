Maude> 
	    Full Maude 2.7 June 18th 2014


	    Logical Model Checker January 10th 2012

Introduced module BAKERY-SATISFACTION1

SymbolicModelCheckerSymbol: examined 1 logical state in bound 1.
SymbolicModelCheckerSymbol: examined 3 logical states in bound 2.
SymbolicModelCheckerSymbol: examined 5 logical states in bound 3.
SymbolicModelCheckerSymbol: examined 9 logical states in bound 4.
SymbolicModelCheckerSymbol: examined 11 logical states in bound 5.
SymbolicModelCheckerSymbol: examined 11 logical states.
logical folding model check in BAKERY-SATISFACTION1 :
  N:Name ; N:Name ;[idle][idle] |= [](ever-wait? -> <> ever-crit?)
result:
  true (complete with depth 5)

SymbolicModelCheckerSymbol: examined 1 logical state in bound 1.
SymbolicModelCheckerSymbol: examined 5 logical states in bound 2.
SymbolicModelCheckerSymbol: examined 9 logical states in bound 3.
SymbolicModelCheckerSymbol: examined 20 logical states in bound 4.
SymbolicModelCheckerSymbol: examined 36 logical states.
logical folding model check in BAKERY-SATISFACTION1 :
  N:Name ; M:Name ; IS:ProcIdleSet |= [](ever-wait? -> <> ever-crit?)
result:
  counterexample found at depth 5
  
prefix
  {(s %5:Name); 0 ; %4:ProcIdleSet,'#11 <- #15:ProcIdleSet ; '#12 <- #14:Name ;
    '#13 <- #15:ProcIdleSet ; '%4 <- #15:ProcIdleSet[idle]; '%5 <- #14:Name,
    'wake}
  {(s s #14:Name); 0 ; #15:ProcIdleSet[wait(s #14:Name)],'#14 <- #16:Name ;
    '#15 <- #16:Name ; '#15 <-[idle],'wake}
loop
  {(s s s #16:Name); 0 ;[wait(s #16:Name)][wait(s s #16:Name)],none,deadlock}  

SymbolicModelCheckerSymbol: examined 1 logical state in bound 1.
SymbolicModelCheckerSymbol: examined 5 logical states in bound 2.
SymbolicModelCheckerSymbol: examined 9 logical states.
logical folding model check in BAKERY-SATISFACTION1 :
  N:Name ; M:Name ; WS:ProcWaitSet |= [](ever-wait? -> <> ever-crit?)
result:
  counterexample found at depth 3
  
prefix
  {(s %5:Name); 0 ; %4:ProcWaitSet,'#11 <- #15:ProcWaitSet ; '#12 <- #14:Name ;
    '#13 <- #15:ProcWaitSet ; '%4 <- #15:ProcWaitSet[idle]; '%5 <- #14:Name,
    'wake}
loop
  {(s %5:Name); 0 ; %4:ProcWaitSet,'#11 <- #15:ProcWaitSet ; '#12 <- #14:Name ;
    '#13 <- #15:ProcWaitSet ; '%4 <- #15:ProcWaitSet[idle]; '%5 <- #14:Name,
    'wake}  

Introduced module BAKERY-SATISFACTION2

SymbolicModelCheckerSymbol: examined 1 logical state in bound 1.
SymbolicModelCheckerSymbol: examined 2 logical states in bound 2.
SymbolicModelCheckerSymbol: examined 3 logical states in bound 3.
SymbolicModelCheckerSymbol: examined 5 logical states in bound 4.
SymbolicModelCheckerSymbol: examined 7 logical states in bound 5.
SymbolicModelCheckerSymbol: examined 8 logical states in bound 6.
SymbolicModelCheckerSymbol: examined 10 logical states in bound 7.
SymbolicModelCheckerSymbol: examined 12 logical states in bound 8.
SymbolicModelCheckerSymbol: examined 13 logical states in bound 9.
SymbolicModelCheckerSymbol: examined 15 logical states.
logical model check in BAKERY-SATISFACTION2 :
  N:Name ; N:Name ;[idle][idle] |= []ex?
result:
  no counterexample found within bound 10

SymbolicModelCheckerSymbol: examined 1 logical state in bound 1.
SymbolicModelCheckerSymbol: examined 2 logical states in bound 2.
SymbolicModelCheckerSymbol: examined 3 logical states in bound 3.
SymbolicModelCheckerSymbol: examined 5 logical states in bound 4.
SymbolicModelCheckerSymbol: examined 6 logical states in bound 5.
SymbolicModelCheckerSymbol: examined 6 logical states.
logical folding model check in BAKERY-SATISFACTION2 :
  N:Name ; N:Name ;[idle][idle] |= []ex?
result:
  true (complete with depth 5)

SymbolicModelCheckerSymbol: examined 1 logical state in bound 1.
SymbolicModelCheckerSymbol: examined 2 logical states in bound 2.
SymbolicModelCheckerSymbol: examined 3 logical states in bound 3.
SymbolicModelCheckerSymbol: examined 5 logical states in bound 4.
SymbolicModelCheckerSymbol: examined 9 logical states.
logical folding model check in BAKERY-SATISFACTION2 :
  N:Name ; M:Name ; WS:ProcWaitSet |= []ex?
result:
  counterexample found at depth 5
  
prefix
  {#1:Name ; #2:Name ; #3:ProcWaitSet,'#10 <- #13:ProcWaitSet ; '#1 <- #11:Name
    ; '#2 <- #12:Name ; '#3 <- #13:ProcWaitSet[wait(#12:Name)],'crit}
  {#11:Name ; #12:Name ; #13:ProcWaitSet[crit(#12:Name)],'#11 <- #13:Name ;
    '#12 <- #14:Name ; '#13 <-[wait(#14:Name)],'crit}
  {#13:Name ; #14:Name ;[crit(#14:Name)][crit(#14:Name)],'#13 <- #15:Name ;
    '#14 <- #16:Name,'exit}
loop
  nil  

SymbolicModelCheckerSymbol: examined 1 logical state in bound 1.
SymbolicModelCheckerSymbol: examined 2 logical states in bound 2.
SymbolicModelCheckerSymbol: examined 6 logical states in bound 3.
SymbolicModelCheckerSymbol: examined 16 logical states in bound 4.
SymbolicModelCheckerSymbol: examined 32 logical states.
logical folding model check in BAKERY-SATISFACTION2 :
  N:Name ; N:Name ; WS:ProcWaitSet |= []ex?
result:
  counterexample found at depth 5
  
prefix
  {#1:Name ; #1:Name ; #2:ProcWaitSet,'#1 <- #9:Name ; '#2 <- #10:ProcWaitSet[
    wait(#9:Name)]; '#8 <- #10:ProcWaitSet,'crit}
  {#9:Name ; #9:Name ; #10:ProcWaitSet[crit(#9:Name)],'#10 <-[wait(#10:Name)];
    '#9 <- #10:Name,'crit}
  {#10:Name ; #10:Name ;[crit(#10:Name)][crit(#10:Name)],'#10 <- #11:Name,
    'exit}
loop
  nil  

Introduced module BAKERY-SATISFACTION2-ABS

SymbolicModelCheckerSymbol: examined 1 logical state in bound 1.
SymbolicModelCheckerSymbol: examined 2 logical states in bound 2.
SymbolicModelCheckerSymbol: examined 4 logical states in bound 3.
SymbolicModelCheckerSymbol: examined 8 logical states in bound 4.
SymbolicModelCheckerSymbol: examined 10 logical states in bound 5.
SymbolicModelCheckerSymbol: examined 10 logical states.
logical folding model check in BAKERY-SATISFACTION2-ABS :
  N:Name ; N:Name ; IS:ProcIdleSet |= []ex?
result:
  true (complete with depth 5)

Maude> Bye.
